### ISSUE TITLE
add support for robust covariance estimators?

### ISSUE THREAD

**Issue by mcreel** [ericqu/LinearRegression.jl#1]
It would be nice to have support for robust covariance estimators. I suggest that the White estimator for cases of heteroskedasticity, and the Newey-West estimator for cases of het. and autocorrelation be considered, as a starting point, as those are probably the most commonly used. It may be possible to make use of the covariance estimators in the packages mentioned on discourse. A possible API could be a cov keyword, defaulting to the ordinary estimator. Something like
`lr = regress(@formula(y ~ 1 + x), df, cov="white")` or 
`lr = regress(@formula(y ~ 1 + x), df, cov="nw")`

**Comment by ericqu** [comment]
I took more time than I hoped. 

- I added several options: "hc0, "hc1, "hc2" and "hc3" following mostly the logic described in as documentation. And tested the results vs the results from SAS (I use SAS for work), for the option "white" it will select hc3 when n < 250 and otherwise select hc0. I hope this makes sense. Please let me know your thoughts.
- I added only one option for "nw" using your code with minor adaptions. I tested with the Nerlove.csv dataset and the results using your ols. Although I reckon that the dataset is not showing some autocorrelation characteristics, do you think that is an issue? do you have a better dataset to test that you could share? I wanted to use the CovarianceMatrices.jl package as it has more options, however, the integration appeared more difficult than expected. I opened an issue there but I don't know how quickly there can be progress there.

so a code like this:
```julia
    cost = [0.082, 0.661, 0.99, 0.315, 0.197, 0.098, 0.949, 0.675, 0.525, 0.501, 1.194, 0.67, 0.349, 0.423, 0.501, 0.55, 0.795, 0.664, 0.705, 0.903, 1.504, 1.615, 1.127, 0.718, 2.414, 1.13, 0.992, 1.554, 1.225, 1.565, 1.936, 3.154, 2.599, 3.298, 2.441, 2.031, 4.666, 1.834, 2.072, 2.039, 3.398, 3.083, 2.344, 2.382, 2.657, 1.705, 3.23, 5.049, 3.814, 4.58, 4.358, 4.714, 4.357, 3.919, 3.442, 4.898, 3.584, 5.535, 4.406, 4.289, 6.731, 6.895, 5.112, 5.141, 5.72, 4.691, 6.832, 4.813, 6.754, 5.127, 6.388, 4.509, 7.185, 6.8, 7.743, 7.968, 8.858, 8.588, 6.449, 8.488, 8.877, 10.274, 6.024, 8.258, 13.376, 10.69, 8.308, 6.082, 9.284, 10.879, 8.477, 6.877, 15.106, 8.031, 8.082, 10.866, 8.596, 8.673, 15.437, 8.211, 11.982, 16.674, 12.62, 12.905, 11.615, 9.321, 12.962, 16.932, 9.648, 18.35, 17.333, 12.015, 11.32, 22.337, 19.035, 12.205, 17.078, 25.528, 24.021, 32.197, 26.652, 20.164, 14.132, 21.41, 23.244, 29.845, 32.318, 21.988, 35.229, 17.467, 22.828, 33.154, 32.228, 34.168, 40.594, 33.354, 64.542, 41.238, 47.993, 69.878, 44.894, 67.12, 73.05, 139.422, 119.939]
    output = [2, 3, 4, 4, 5, 9, 11, 13, 13, 22, 25, 25, 35, 39, 43, 63, 68, 81, 84, 73, 99, 101, 119, 120, 122, 130, 138, 149, 196, 197, 209, 214, 220, 234, 235, 253, 279, 290, 290, 295, 299, 324, 333, 338, 353, 353, 416, 420, 456, 484, 516, 550, 563, 566, 592, 671, 696, 719, 742, 795, 800, 808, 811, 855, 860, 909, 913, 924, 984, 991, 1000, 1098, 1109, 1118, 1122, 1137, 1156, 1166, 1170, 1215, 1279, 1291, 1290, 1331, 1373, 1420, 1474, 1497, 1545, 1649, 1668, 1782, 1831, 1833, 1838, 1787, 1918, 1930, 2028, 2057, 2084, 2226, 2304, 2341, 2353, 2367, 2451, 2457, 2507, 2530, 2576, 2607, 2870, 2993, 3202, 3286, 3312, 3498, 3538, 3794, 3841, 4014, 4217, 4305, 4494, 4764, 5277, 5283, 5668, 5681, 5819, 6000, 6119, 6136, 7193, 7886, 8419, 8642, 8787, 9484, 9956, 11477, 11796, 14359, 16719]
    labor = [2.09, 2.05, 2.05, 1.83, 2.12, 2.12, 1.98, 2.05, 2.19, 1.72, 2.09, 1.68, 1.81, 2.3, 1.75, 1.76, 1.98, 2.29, 2.19, 1.75, 2.2, 1.66, 1.92, 1.77, 2.09, 1.82, 1.8, 1.92, 1.92, 2.19, 1.92, 1.52, 1.92, 2.2, 2.11, 1.92, 2.05, 1.66, 1.8, 1.77, 1.7, 2.05, 2.19, 1.85, 2.19, 2.13, 1.54, 1.52, 2.09, 1.75, 2.3, 2.05, 2.32, 2.31, 1.92, 2.05, 1.76, 1.7, 2.04, 2.24, 1.7, 1.68, 2.29, 2.0, 2.31, 1.45, 1.7, 1.76, 1.7, 2.09, 1.55, 2.11, 2.05, 2.3, 2.19, 2.04, 2.31, 1.7, 2.05, 2.19, 2.0, 2.32, 1.55, 2.13, 2.2, 2.2, 1.85, 1.76, 1.8, 2.32, 1.8, 2.13, 1.98, 1.76, 1.45, 2.24, 1.69, 1.81, 2.11, 1.76, 1.77, 2.0, 2.3, 2.04, 1.69, 1.76, 2.04, 2.2, 1.76, 2.31, 1.92, 1.76, 1.76, 2.31, 2.3, 1.61, 1.68, 2.09, 2.09, 2.05, 2.29, 2.11, 1.53, 2.11, 2.04, 2.19, 1.92, 2.04, 2.11, 1.76, 1.79, 2.11, 1.54, 1.92, 2.12, 1.61, 2.32, 2.24, 2.31, 2.11, 1.68, 2.24, 2.12, 2.31, 2.3]
    fuel = [17.9, 35.1, 35.1, 32.2, 28.6, 28.6, 35.5, 35.1, 29.1, 15.0, 17.9, 39.7, 22.6, 23.6, 42.8, 10.3, 35.5, 28.5, 29.1, 42.8, 36.2, 33.4, 22.5, 21.3, 17.9, 38.9, 20.2, 22.5, 29.1, 29.1, 22.5, 27.5, 22.5, 36.2, 24.4, 22.5, 35.1, 33.4, 20.2, 21.3, 26.9, 35.1, 29.1, 24.6, 29.1, 10.7, 26.2, 27.5, 30.0, 42.8, 23.6, 35.1, 31.9, 33.5, 22.5, 35.1, 10.3, 26.9, 20.7, 26.5, 26.9, 39.7, 28.5, 34.3, 33.5, 17.6, 26.9, 10.3, 26.9, 30.0, 28.2, 24.4, 35.1, 23.6, 29.1, 20.7, 33.5, 26.9, 35.1, 29.1, 34.3, 31.9, 28.2, 30.0, 36.2, 36.2, 24.6, 10.3, 20.2, 31.9, 20.2, 10.7, 35.5, 10.3, 17.6, 26.5, 12.9, 22.6, 24.4, 10.3, 21.3, 34.3, 23.6, 20.7, 12.9, 10.3, 20.7, 36.2, 10.3, 33.5, 22.5, 10.3, 10.3, 33.5, 23.6, 17.8, 28.8, 30.0, 30.0, 35.1, 28.5, 24.4, 18.1, 24.4, 20.7, 29.1, 29.1, 20.7, 24.4, 10.3, 18.5, 24.4, 26.2, 22.5, 28.6, 17.8, 31.9, 26.5, 33.5, 24.4, 28.8, 26.5, 28.6, 33.5, 23.6]
    capital = [183, 174, 171, 166, 233, 195, 206, 150, 155, 188, 170, 167, 213, 164, 170, 161, 210, 158, 156, 176, 170, 192, 164, 175, 180, 176, 202, 227, 186, 183, 169, 168, 164, 164, 170, 158, 177, 195, 176, 188, 187, 152, 157, 163, 143, 167, 217, 144, 178, 176, 167, 158, 162, 198, 164, 164, 161, 174, 157, 185, 157, 203, 178, 183, 168, 196, 166, 172, 158, 174, 225, 168, 177, 161, 162, 158, 176, 183, 166, 164, 207, 175, 225, 178, 157, 138, 163, 168, 158, 177, 170, 183, 162, 177, 196, 164, 158, 157, 163, 161, 156, 217, 161, 183, 167, 161, 163, 170, 174, 197, 162, 155, 167, 176, 170, 183, 190, 170, 176, 159, 157, 161, 172, 203, 167, 195, 161, 159, 177, 157, 196, 183, 189, 160, 162, 178, 199, 182, 190, 165, 203, 151, 148, 212, 162]
    ndf = DataFrame(cost = cost, output = output, labor = labor, fuel = fuel, capital = capital)

    f = @formula(log(cost) ~ 1 + log(output) + log(labor) + log(fuel) + log(capital))
    lr2 = regress(f, ndf, cov=[:hc0, :nw])
```
would give results like this:
```
Model definition:       :(log(cost)) ~ 1 + :(log(output)) + :(log(labor)) + :(log(fuel)) + :(log(capital))
Used observations:      145
Model statistics:
  R²: 0.925955                  Adjusted R²: 0.92384
  MSE: 0.153943                 RMSE: 0.392356
  σ̂²: 0.153943                  AIC: -266.408
Confidence interval: 95%
Coefficients statistics:
Terms ╲ Stats │       Coefs      Std err            t     Pr(>|t|)       low ci      high ci          VIF
──────────────┼──────────────────────────────────────────────────────────────────────────────────────────
(Intercept)   │     -3.5265      1.77437     -1.98747    0.0488212     -7.03452   -0.0184845          0.0
log(output)   │    0.720394    0.0174664      41.2445  3.29949e-80     0.685862     0.754926      1.04411
log(labor)    │    0.436341     0.291048      1.49921     0.136071    -0.139076      1.01176       1.2058
log(fuel)     │    0.426517     0.100369      4.24948   3.88881e-5     0.228082     0.624952      1.21381
log(capital)  │   -0.219888     0.339429    -0.647819     0.518163    -0.890957      0.45118      1.08612

White's covariance estimator (HC0):
Terms ╲ Stats │       Coefs      Std err            t     Pr(>|t|)
──────────────┼───────────────────────────────────────────────────
(Intercept)   │     -3.5265      1.68871     -2.08828    0.0385835
log(output)   │    0.720394    0.0320306      22.4908  2.51567e-48
log(labor)    │    0.436341     0.241364      1.80782    0.0727816
log(fuel)     │    0.426517    0.0741699      5.75054   5.35856e-8
log(capital)  │   -0.219888      0.31818    -0.691081     0.490659

Newey-West's covariance estimator:
Terms ╲ Stats │       Coefs      Std err            t     Pr(>|t|)
──────────────┼───────────────────────────────────────────────────
(Intercept)   │     -3.5265      1.54133     -2.28796    0.0236382
log(output)   │    0.720394    0.0455834      15.8039  6.25242e-33
log(labor)    │    0.436341     0.250279      1.74342    0.0834553
log(fuel)     │    0.426517    0.0728964        5.851   3.30623e-8
log(capital)  │   -0.219888     0.265114    -0.829411     0.408284
```

I thought it was interesting to keep the initial Std Err to contrast them with the ones coming from White/Newey-west estimators. I hope this is alright.

Please let me know your thoughts.

**Comment by ericqu** [comment]
and to illustrate the "white"parameter:
```julia
lr2 = regress(f, ndf, cov=[:white])
```
gives :
```julia
Model definition:       :(log(cost)) ~ 1 + :(log(output)) + :(log(labor)) + :(log(fuel)) + :(log(capital))
Used observations:      145
Model statistics:
  R²: 0.925955                  Adjusted R²: 0.92384
  MSE: 0.153943                 RMSE: 0.392356
  σ̂²: 0.153943                  AIC: -266.408
Confidence interval: 95%
Coefficients statistics:
Terms ╲ Stats │       Coefs      Std err            t     Pr(>|t|)       low ci      high ci          VIF
──────────────┼──────────────────────────────────────────────────────────────────────────────────────────
(Intercept)   │     -3.5265      1.77437     -1.98747    0.0488212     -7.03452   -0.0184845          0.0
log(output)   │    0.720394    0.0174664      41.2445  3.29949e-80     0.685862     0.754926      1.04411
log(labor)    │    0.436341     0.291048      1.49921     0.136071    -0.139076      1.01176       1.2058
log(fuel)     │    0.426517     0.100369      4.24948   3.88881e-5     0.228082     0.624952      1.21381
log(capital)  │   -0.219888     0.339429    -0.647819     0.518163    -0.890957      0.45118      1.08612

White's covariance estimator (HC3):
Terms ╲ Stats │       Coefs      Std err            t     Pr(>|t|)
──────────────┼───────────────────────────────────────────────────
(Intercept)   │     -3.5265        1.908     -1.84828    0.0666724
log(output)   │    0.720394    0.0362112      19.8942  1.23784e-42
log(labor)    │    0.436341     0.267632      1.63038     0.105269
log(fuel)     │    0.426517    0.0821241      5.19356   7.12877e-7
log(capital)  │   -0.219888     0.358333    -0.613642     0.540447
```

**Comment by mcreel** [comment]
This looks very nice to me, I like the option to use more than one covariance estimator.  Thanks for this enhancement, I will probably start using this package in my teaching when I have a chance to update.

A couple of suggestions:

1) It would be nice if the confidence intervals could be computed using the robust standard errors, too.

2) I would consider some way of offering only the output for the versions that are selected, and suppressing the plain OLS output, In many cases, in applied econometrics, if a person is working with cross sectional data, they will always want heteroskedasticity robust standard errors, and will not be interested in the plain vanilla OLS version. Likewise, with time series data, the NW version, or something similar, is what one would pay attention to. Too much output for extraneous covariance estimators could be a distraction.
3. I'm not sure that sigma^2 and AIC are worth reporting when robust covariance estimators are used, because the assumption is that the errors may not be IID.

Regarding a data set where autocorrelation is a problem, I will look for something.

Thanks!

**Comment by ericqu** [comment]
Thank you for the feedback.
Given your feedback about the multiple covariance estimators, I am not sure I implemented it in the most useful way. As of now, if the user selects "hc0" and "hc1", I will take only the first one from all the heteroskedasticity keywords, and similarly, it will take one from the HAC keywords (when there will be more than one). However, I now wonder if it is more useful to take all the keywords provided by the user? That way, the user sees both hc0 and hc1 and could compare them. Right?

1. The confidence interval can be computed; I didn't find information about computing them. I suppose I can compute them the same way as the regular OLS confidence intervals, or is there a specific way?

2. Good to know! I will look into not displaying the regular OLS when robust covariance is requested.

3. Makes sense. I will look into not displaying this information when robust covariance is requested. This leads to another question; the sigma^2 is used to compute the standardized and studentized residuals, the confidence intervals for the predictions and the prediction means, and the Cook's Distance. I wonder if other values should be computed instead that I could use to compute these stats? Or maybe indicating as a warning that because the assumptions behind the Sigma^2 are invalid, these stats should be used with caution?

**Comment by mcreel** [comment]
I think that presenting results for all selected options would be nice. The default should probably be vanilla OLS. If the user selects something else, then I suggest presenting only those results, without vanilla OLS. There could be a way to request both vanilla and something else. About the numbered points:
1. Confidence intervals can be computed in exactly the same way as is done using vanilla OLS, you just replace the estimated standard errors of the coefficients with the robust version. The theory that justifies the results is now asymptotic and relies on the CLT, rather than small sample with a normality assumption. The are variations that are possible, like taking critical values from the normal distribution, rather than the Student-t.
2. OK, that's what I commented on above.
3. I think that standardized residuals will not be possible when using robust covariances, as these are used when there is heteroscedasticity and/or autocorrelation of _unknown_ form. If the forms were known, then one could do a generalized least squares tranformation to get back to vanilla OLS. Probably adding a warning for these stats is the easiest option.

Regarding an example of data with autocorrelation, the data https://github.com/mcreel/Econometrics/blob/master/Examples/Data/sp500.csv is on returns for the SP500 index. If you regress squared rets on two lags of squared rets, the errors will be strongly autocorrelated. The regression I have run is at https://github.com/mcreel/Econometrics/blob/master/Examples/TimeSeries/SP500.jl

**Comment by ericqu** [comment]
The features are now added in the latest version (0.71). 
I added the test for Newey-West and got the same results as with your OLS.
The documentation is missing as I am still looking into an efficient way to have the documentation.
Nevertheless the ```lr2 = regress(f, ndf, cov=["hc0", "hc1", "nw"])``` will give you the HC0, HC1 and Newey-West results.
The tests give some additional examples of how to use it.
